packages/server/src/tests/banners/bannerSelection.ts

```diff
@@ -96,6 +96,25 @@ const getForcedVariant = (
     return null;
 };
 
+const purchaseMatches = (
+    test: BannerTest,
+    purchaseInfo: BannerTargeting['purchaseInfo'],
+    isSignedIn: boolean,
+) => {
+    const { purchaseInfo: testPurchaseInfo } = test;
+
+    // Ignore tests specifying purchase info if user is signed in / if no purchase info in targeting
+    if (isSignedIn || !purchaseInfo) {
+        return !testPurchaseInfo;
+    }
+
+    const { product, userType } = purchaseInfo;
+    const productValid = product && testPurchaseInfo?.product.includes(product);
+    const userValid = userType && testPurchaseInfo?.userType.includes(userType);
+
+    return productValid && userValid;
+};
```

Comment:
I should probably spin this out into a separate util, but I think I could leave that for the gate?